packages/flutter/lib/src/material/banner.dart

```diff
@@ -138,18 +179,109 @@ class MaterialBanner extends StatelessWidget {
   /// Defaults to [OverflowBarAlignment.end].
   final OverflowBarAlignment overflowAlignment;
 
+  /// The animation driving the entrance and exit of the material banner.
+  final Animation<double>? animation;
+
+  /// Called the first time that the material banner is visible within a [Scaffold].
+  final VoidCallback? onVisible;
+
+  // API for ScaffoldMessengerState.showMaterialBanner():
+
+  /// Creates an animation controller useful for driving a material banner's entrance and exit animation.
+  static AnimationController createAnimationController({ required TickerProvider vsync }) {
+    return AnimationController(
+      duration: _materialBannerTransitionDuration,
+      debugLabel: 'MaterialBanner',
+      vsync: vsync,
+    );
+  }
+
+  /// Creates a copy of this snack bar but with the animation replaced with the given animation.
+  ///
+  /// If the original snack bar lacks a key, the newly created snack bar will
+  /// use the given fallback key.
+  MaterialBanner withAnimation(Animation<double> newAnimation, { Key? fallbackKey }) {
+    return MaterialBanner(
+      key: key ?? fallbackKey,
+      content: content,
+      contentTextStyle: contentTextStyle,
+      actions: actions,
+      leading: leading,
+      backgroundColor: backgroundColor,
+      padding: padding,
+      leadingPadding: leadingPadding,
+      forceActionsBelow: forceActionsBelow,
+      overflowAlignment: overflowAlignment,
+      animation: newAnimation,
+      onVisible: onVisible,
+    );
+  }
+
+  @override
+  State<MaterialBanner> createState() => _MaterialBannerState();
+}
+
+class _MaterialBannerState extends State<MaterialBanner> with TickerProviderStateMixin {
+  // TODO(Calamity210): Remove the private animation once MaterialBanners have migrated to the new API.
+  // A mutable animation is need to allow us to create a fallback animation for backwards compatibility
+  // during the opt-in phase of MaterialBanner's migration to the ScaffoldMessenger API.
+  Animation<double>? _animation;
```

Comment:
Instead of this, if the animation is null, can we just make it a static banner? That way it will still be backwards compatible.

Post-migration, we you planning to have the animation value required? or will it always be generated for the user? I am curious about the comment here.

Reply:
> @Piinks You may want to take a look at [this commit](https://github.com/flutter/flutter/commit/8cda0f3c0ebae4af70eba803ee3b65875f4dd655) while running those tests. I had added it to refrain from breaking existing code during the "opt-in" period of migrating to the changes within this PR.

Right, I see that now. I think that the fallback animation is helpful, but still breaks users because just creating a banner shows nothing, since the animation controller is not being driven by the ScaffoldMessenger - maybe set the animation value to 1 if the fallback is used? Will that work?

Reply:
Sure! No rush at all. :)